statistics/row_sampler.go

```diff
@@ -250,28 +269,41 @@ func (s *baseCollector) ToProto() *tipb.RowSampleCollector {
 	return collector
 }
 
-func (s *baseCollector) FromProto(pbCollector *tipb.RowSampleCollector) {
+func (s *baseCollector) FromProto(pbCollector *tipb.RowSampleCollector, memTracker *memory.Tracker) {
 	s.Count = pbCollector.Count
 	s.NullCount = pbCollector.NullCounts
 	s.FMSketches = make([]*FMSketch, 0, len(pbCollector.FmSketch))
 	for _, pbSketch := range pbCollector.FmSketch {
 		s.FMSketches = append(s.FMSketches, FMSketchFromProto(pbSketch))
 	}
 	s.TotalSizes = pbCollector.TotalSize
-	s.Samples = make(WeightedRowSampleHeap, 0, len(pbCollector.Samples))
+	sampleNum := len(pbCollector.Samples)
+	s.Samples = make(WeightedRowSampleHeap, 0, sampleNum)
+	// consume mandatory memory at the beginning, including all empty ReservoirRowSampleItems and all empty Datums of all sample rows, if exceeds, fast fail
+	if len(pbCollector.Samples) > 0 {
+		rowLen := len(pbCollector.Samples[0].Row)
+		// 8 is the size of reference
+		initMemSize := int64(sampleNum) * (int64(rowLen)*types.EmptyDatumSize + EmptyReservoirSampleItemSize + 8)
+		s.MemSize += initMemSize
+		memTracker.Consume(initMemSize)
+	}
+	bufferedMemSize := int64(0)
 	for _, pbSample := range pbCollector.Samples {
-		data := make([]types.Datum, 0, len(pbSample.Row))
+		rowLen := len(pbSample.Row)
+		data := make([]types.Datum, 0, rowLen)
 		for _, col := range pbSample.Row {
 			b := make([]byte, len(col))
 			copy(b, col)
 			data = append(data, types.NewBytesDatum(b))
 		}
 		// Directly copy the weight.
-		s.Samples = append(s.Samples, &ReservoirRowSampleItem{
-			Columns: data,
-			Weight:  pbSample.Weight,
-		})
+		sampleItem := &ReservoirRowSampleItem{Columns: data, Weight: pbSample.Weight}
+		s.Samples = append(s.Samples, sampleItem)
+		deltaSize := sampleItem.MemUsage() - EmptyReservoirSampleItemSize - int64(rowLen)*types.EmptyDatumSize
+		memTracker.BufferedConsume(&bufferedMemSize, deltaSize)
+		s.MemSize += deltaSize
```

Comment:
Why not sum the bufferedMemSize and Consume once out of the loop?

Reply:
ditto